File: GfG.py
class Node:
    count = 0

    def __init__(self, nim, nama):
        self.nim = nim
        self.nama = nama
        self.next = None
        Node.count += 1


# Insert at beginning
def insert_at_front(head, nim, nama):
    new_node = Node(nim, nama)
    new_node.next = head
    return new_node


# Insert at given position
def insert_pos(head, pos, nim, nama):

    if pos < 1 or pos > Node.count + 1:
        return head

    if pos == 1:
        return insert_at_front(head, nim, nama)

    curr = head
    for _ in range(pos - 2):
        curr = curr.next

    new_node = Node(nim, nama)
    new_node.next = curr.next
    curr.next = new_node

    return head


# Insert at end
def insert_at_end(head, nim, nama):

    new_node = Node(nim, nama)

    if head is None:
        return new_node

    last = head
    while last.next is not None:
        last = last.next

    last.next = new_node
    return head


# Delete from beginning
def delete_head(head):

    if head is None:
        return None

    head = head.next
    Node.count -= 1
    return head


# Delete given position
def delete_pos(head, pos):

    if head is None or pos < 1 or pos > Node.count:
        return head

    if pos == 1:
        return delete_head(head)

    curr = head
    prev = None

    for _ in range(pos - 1):
        prev = curr
        curr = curr.next

    prev.next = curr.next
    Node.count -= 1
    return head


# Delete from end
def delete_end(head):

    if head is None:
        return None

    if head.next is None:
        Node.count -= 1
        return None

    curr = head
    while curr.next.next is not None:
        curr = curr.next

    curr.next = None
    Node.count -= 1
    return head


# Delete first occurrence by NIM
def delete_by_nim(head, nim):

    if head is None:
        return None

    if head.nim == nim:
        Node.count -= 1
        return head.next

    curr = head
    while curr.next is not None and curr.next.nim != nim:
        curr = curr.next

    if curr.next is not None:
        curr.next = curr.next.next
        Node.count -= 1

    return head


# Show data
def show_data(head):

    if head is None:
        print("Data kosong")
        return

    curr = head
    while curr is not None:
        print("NIM :", curr.nim, "| Nama :", curr.nama)
        curr = curr.next


# MAIN PROGRAM
head = None

while True:

    print("\n===== MENU LINKED LIST =====")
    print("1. Insert at beginning")
    print("2. Insert at given position")
    print("3. Insert at end")
    print("4. Delete from beginning")
    print("5. Delete given position")
    print("6. Delete from end")
    print("7. Delete first occurence (by NIM)")
    print("8. Show data")
    print("9. Exit")

    pilihan = int(input("Pilih: "))

    if pilihan == 1:
        nim = input("Masukkan NIM: ")
        nama = input("Masukkan Nama: ")
        head = insert_at_front(head, nim, nama)

    elif pilihan == 2:
        pos = int(input("Masukkan posisi: "))
        nim = input("Masukkan NIM: ")
        nama = input("Masukkan Nama: ")
        head = insert_pos(head, pos, nim, nama)

    elif pilihan == 3:
        nim = input("Masukkan NIM: ")
        nama = input("Masukkan Nama: ")
        head = insert_at_end(head, nim, nama)

    elif pilihan == 4:
        head = delete_head(head)

    elif pilihan == 5:
        pos = int(input("Masukkan posisi delete: "))
        head = delete_pos(head, pos)

    elif pilihan == 6:
        head = delete_end(head)

    elif pilihan == 7:
        nim = input("Masukkan NIM yang dihapus: ")
        head = delete_by_nim(head, nim)

    elif pilihan == 8:
        show_data(head)
        print("Jumlah data =", Node.count)

    elif pilihan == 9:
        print("Program selesai")
        break

    else:
        print("Pilihan tidak valid")

File: GfG.java
import java.util.Scanner;

class Node {
    String nim;
    String nama;
    Node next;
    static int count = 0;

    Node(String nim, String nama) {
        this.nim = nim;
        this.nama = nama;
        this.next = null;
        count++; // insert otomatis tambah count
    }
}

public class GfG {

    // Insert at beginning
    static Node insertAtFront(Node head, String nim, String nama) {
        Node newNode = new Node(nim, nama);
        newNode.next = head;
        return newNode;
    }

    // Insert at given position
    static Node insertPos(Node head, int pos, String nim, String nama) {

        if (pos < 1 || pos > Node.count + 1)
            return head;

        if (pos == 1) {
            Node newNode = new Node(nim, nama);
            newNode.next = head;
            return newNode;
        }

        Node curr = head;

        for (int i = 1; i < pos - 1; i++)
            curr = curr.next;

        Node newNode = new Node(nim, nama);
        newNode.next = curr.next;
        curr.next = newNode;

        return head;
    }

    // Insert at end
    static Node insertAtEnd(Node head, String nim, String nama) {

        Node newNode = new Node(nim, nama);

        if (head == null)
            return newNode;

        Node last = head;

        while (last.next != null)
            last = last.next;

        last.next = newNode;
        return head;
    }

    // Delete from beginning
    static Node deleteHead(Node head) {

        if (head == null)
            return null;

        head = head.next;
        Node.count--;
        return head;
    }

    // Delete given position
    static Node deleteNode(Node head, int position) {

        if (head == null)
            return null;

        if (position < 1 || position > Node.count)
            return head;

        if (position == 1) {
            head = head.next;
            Node.count--;
            return head;
        }

        Node temp = head;
        Node prev = null;

        for (int i = 1; i < position; i++) {
            prev = temp;
            temp = temp.next;
        }

        prev.next = temp.next;
        Node.count--;
        return head;
    }

    // Delete from end
    static Node removeLastNode(Node head) {

        if (head == null)
            return null;

        if (head.next == null) {
            Node.count--;
            return null;
        }

        Node secondLast = head;

        while (secondLast.next.next != null)
            secondLast = secondLast.next;

        secondLast.next = null;
        Node.count--;
        return head;
    }

    // Delete first occurrence by NIM
    static Node deleteByNim(Node head, String nim) {

        if (head == null)
            return null;

        if (head.nim.equals(nim)) {
            head = head.next;
            Node.count--;
            return head;
        }

        Node curr = head;

        while (curr.next != null && !curr.next.nim.equals(nim))
            curr = curr.next;

        if (curr.next != null) {
            curr.next = curr.next.next;
            Node.count--;
        }

        return head;
    }

    // Show data
    static void printList(Node head) {

        if (head == null) {
            System.out.println("Data kosong");
            return;
        }

        Node curr = head;

        while (curr != null) {
            System.out.println("NIM : " + curr.nim + " | Nama : " + curr.nama);
            curr = curr.next;
        }
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        Node head = null;
        int pilihan;

        do {
            System.out.println("\n===== MENU LINKED LIST =====");
            System.out.println("1. Insert at beginning");
            System.out.println("2. Insert at given position");
            System.out.println("3. Insert at end");
            System.out.println("4. Delete from beginning");
            System.out.println("5. Delete given position");
            System.out.println("6. Delete from end");
            System.out.println("7. Delete first occurence (by NIM)");
            System.out.println("8. Show data");
            System.out.println("9. Exit");
            System.out.print("Pilih: ");
            pilihan = sc.nextInt();
            sc.nextLine();

            switch (pilihan) {

                case 1:
                    System.out.print("Masukkan NIM: ");
                    String nim1 = sc.nextLine();
                    System.out.print("Masukkan Nama: ");
                    String nama1 = sc.nextLine();
                    head = insertAtFront(head, nim1, nama1);
                    break;

                case 2:
                    System.out.print("Masukkan posisi: ");
                    int pos = sc.nextInt();
                    sc.nextLine();
                    System.out.print("Masukkan NIM: ");
                    String nim2 = sc.nextLine();
                    System.out.print("Masukkan Nama: ");
                    String nama2 = sc.nextLine();
                    head = insertPos(head, pos, nim2, nama2);
                    break;

                case 3:
                    System.out.print("Masukkan NIM: ");
                    String nim3 = sc.nextLine();
                    System.out.print("Masukkan Nama: ");
                    String nama3 = sc.nextLine();
                    head = insertAtEnd(head, nim3, nama3);
                    break;

                case 4:
                    head = deleteHead(head);
                    break;

                case 5:
                    System.out.print("Masukkan posisi delete: ");
                    int delPos = sc.nextInt();
                    head = deleteNode(head, delPos);
                    break;

                case 6:
                    head = removeLastNode(head);
                    break;

                case 7:
                    System.out.print("Masukkan NIM yang dihapus: ");
                    String delNim = sc.nextLine();
                    head = deleteByNim(head, delNim);
                    break;

                case 8:
                    printList(head);
                    System.out.println("Jumlah data = " + Node.count);
                    break;

                case 9:
                    System.out.println("Program selesai");
                    break;

                default:
                    System.out.println("Pilihan tidak ada");
            }

        } while (pilihan != 9);
    }
}
